# Entry	 point to generate R report for OxoG Filter version 3
# 
# IMPORTANT NOTE:  This script assumes the naming convention of the figures in the OxoG filter.  
#  If that convention changes, figures will no longer display properly. 
#
# Author: lichtens
###############################################################################
#main <- function(...) {
	## Arguments
	argCaseFile = NULL;
	argFigureListFile = NULL;
	usage <- paste(sep="","\r\n", "Usage: createReport.R -LnozzlePath -CcaseListFile -Ffigurefile -TtableSummaryFile -MfilteredMafFile -RmafFile\r\n",
			"nozzlePath is where the Nozzle R package can be found.\r\n",
			"caseListFile is the list of cases.\r\n",
			"figureFile is a file containing a list of case to figures file mapping\r\n",
			"tableSummaryFile is the location of the summary table generated by the OxoG v3 filter \r\n",
			"filteredMafFile is the location of the resulting filtered maf table generated by the OxoG v3 filter \r\n",
			"mafFile is the location of the resulting unfiltered maf table generated by the OxoG v3 filter \r\n",
			"\r\n",
			"$HeadURL: https://svn.broadinstitute.org/CancerGenomeAnalysis/trunk/matlab/seq/oxoGFilterv3NozzleReport/tags/v1.10.2.0/createReport.R $"
			);
	args = commandArgs();

	minArgs = 6;
	if (length(args) < minArgs){
		cat(paste (sep=" ","Only",length(args), " arguments found.  Need", minArgs,"." ));
		stop(usage);
	}
	nozzlePath = ""
	for(i in 1:length(args)) {
		flag = substring(args[[i]], 0, 2)
		value = substring(args[[i]], 3)
		if (flag == '-C') {
			caseListFile=value 
		} else if(flag=='-F') {
			figureFile=value
		} else if(flag=='-T') {
			tableSummaryFile=value
		} else if(flag=='-M') {
			filteredMafFile=value
		} else if(flag=='-R') {
			mafFile=value
		} else if(flag=='-L') {
			nozzlePath=value
		} 
	}
	
	#TODO: Rename figureFile and name it as you would a directory
	#TODO: Need more intelligent errors when parameters are missing.
	if (nozzlePath == "") {
		require( "Nozzle.R1" )
	} else {
		require( "Nozzle.R1", lib.loc=nozzlePath, quietly=TRUE);
	}
	oxoReport <- newReport( "OxoG Report" );
	
	oxoReport <- addToIntroduction( oxoReport ,
			newParagraph("Experiments carried out by the sequencing platform have established that the CCG>CAG artifact arises from an oxidation process in library construction. Prior to the PCR step, conversion of guanine to 8-oxoguanine (8-oxoG) tends to occur in the context of CGG (where the 8-oxoG is the middle G) sequences. Unlike regular guanine, 8-oxoG has a higher chemical affinity to bind with adenine rather than cytosine. The presence of 8-oxoG during subsequent PCR amplification cycles introduces adenine bases into DNA molecules at sites where cytosine should have been, systematically producing CAG sequences where the original sequence was CCG. Unlike natural mutations, non-reference artifact bases are locked specific strand orientations by the forked adapter ligation step before PCR, resulting in G>T artifacts on the F1R2 orientation and C>A artifacts on the F2R1 orientation." ),
			newParagraph("Artifact mode mutations are ones which are C>A or G>T.  The oxoG artifacts never manifest with a different signature.")
			);
	
	# copyDir is local dir
	copyDir = ("./");
	if (file.exists(copyDir) == FALSE){
		dir.create(copyDir); 
	} 
	
	# Retrieve the list (in a file) of cases for this report
	if (file.exists(caseListFile) == TRUE) {
		if (file.info(caseListFile)$size == 0 ){
			oxoReport <- addToSummary(oxoReport,
					newParagraph(paste(sep="", "There were no cases in this run."))
			);
			writeReport( oxoReport , filename="nozzle", level=PROTECTION.PUBLIC );	
			quit("no",0)
		} else {
			destCaseFile = paste(copyDir, "allCases.txt", sep="");
			cat(paste(sep=" ", "Copying caseList file... ", caseListFile,"to", destCaseFile, "\r\n"));
			file.copy(caseListFile, destCaseFile);
			case_table = read.table(destCaseFile, header = FALSE);
		}
	} else {
		stop(paste('Could not find file containing case list:', caseListFile, sep=" "));
	}
	
	#TODO: Break into a function
	# Retrieve the list (in a file) of cases for this report
	if (file.exists(tableSummaryFile) == TRUE) {
		cat("Copying table summary file...\r\n")
		destTableSummaryFile = paste(copyDir,basename(tableSummaryFile), sep="");
		cat(tableSummaryFile)
		file.copy(tableSummaryFile, destTableSummaryFile, overwrite=TRUE);
		summaryTableObjTmp = read.table(destTableSummaryFile, header = TRUE);
	} else {
		stop(paste('Could not find file containing table summary:', destTableSummaryFile, sep=" "));
	}
	
	summaryTableObj = summaryTableObjTmp;
	
	# Prepare the data for proper formatting.
	for (i in array(1:length(summaryTableObj[1,])) ) {
		for (j in array(1:length(summaryTableObj[,1])) ) {
			if (is.nan(summaryTableObj[j,i])) {
				cat("Ignoring NaN")		
			} else {
				if (typeof(summaryTableObj[j,i]) == "double"){
					if (summaryTableObj[j,i] >= 1000) {
						summaryTableObj[j,i] = round(summaryTableObj[j,i],0)
					} else if (summaryTableObj[j,i] >= 1.0) {
						summaryTableObj[j,i] = round(summaryTableObj[j,i],2)
					} else {
						summaryTableObj[j,i] = round(summaryTableObj[j,i],4)
					}	
				}
			}
		}
	}
	
	#destFigureDir = paste(copyDir,basename(figureFile),sep="");
	#if (!file.exists(destFigureDir)){
	#	dir.create(destFigureDir); 
	#}
	destFigureDir="."
	
	if (file.exists(figureFile) == TRUE) {
		filenames <- dir(figureFile)
		for (j in 1:length(filenames)){
			cat(filenames[j], '\n')
			file.copy(from=paste(figureFile, '/', filenames[j], sep=''), to=paste(destFigureDir, '/', filenames[j], sep=''), overwrite = TRUE)
		}
				
	} else {
		stop(paste('Could not find dir containing case figures:', figureFile, sep=" "));
	}
	
	# Read the cases from the file.
	caseLines = readLines(caseListFile)
	
	## SUMMARY
	oxoReport <- addToSummary(oxoReport,
			newParagraph(paste(sep="", length(caseLines), " case(s) were processed in this run."))
			);
	
	
	## RESULTS (incl. FIGURES)
	# Create and insert table
	summaryTable <- newTable(summaryTableObj, "Counts and estimates for all cases in the given maf file", file=destTableSummaryFile, significantDigits=8);
			
	# Display summary table
	oxoReport <- addToResults(oxoReport, summaryTable);
	oxoReport <- addToResults(oxoReport, newParagraph(
			paste(sep=" ", 
					"<b>Notes:</b>",
					"<br>case -- case sample (usually Tumor)",
					"<br>control -- control sample (usually Normal)",
					"<br>N -- number of called SNP mutations",
					"<br>N_A -- number of called mutations in the artifact mode (C>A or G>T)",
					"<br>N_nA  -- number of called mutations not in the artifact mode (not C>A nor G>T)",
					"<br>N_filtered -- number of mutations filtered",
					"<br>N_oxo -- estimated number of artifact mutations",
					"<br>N_oxo_low -- low value of 95% confidence interval",
					"<br>N_oxo_hi -- high value of 95% confidence interval",
					"<br>N_mut -- estimated number of real, artifact mode mutations",
					"<br>RR -- estimated number of real rejected mutations.  (i.e. real mutations that were filtered) ",
					"<br>RR_CI_low -- low value of 95% confidence interval",
					"<br>RR_CI_hi -- high value of 95% confidence interval",
					"<br>PA -- estimated number of passing artifacts.  (i.e. artifact SNVs that were not filtered)",
					"<br>PA_CI_low -- low value of 95% confidence interval",
					"<br>PA_CI_hi -- high value of 95% confidence interval",
					"<br>PoxoG -- estimated PoxoG from this sample.  Note that this is not the value used in the filter and is included for informational purposes and possible use in future functionality.  Do not use in any analyses.", 
					"<br>PoxoG_Est_CI_low -- low value of 95% confidence interval",
					"<br>PoxoG_Est_CI_hi -- high value of 95% confidence interval",
					"<br>lod0 -- log odds ratio between the oxoG contamination estimate and zero (no OxoG contamination).  Note that this value is not used and is provided for information only."
				)
			)
	);
	
	# Display MAF Files
	# TODO: These links need to be fixed.
	mafFileSubSection = newSubSection("MAF Files");
	mafFileSubSection <- addTo(mafFileSubSection, newParagraph(paste(sep="","Filtered: ", filteredMafFile )));
	mafFileSubSection <- addTo(mafFileSubSection, newParagraph(paste(sep="","Unfiltered: ", mafFile )));
	oxoReport <- addToResults(oxoReport, mafFileSubSection);
	
	# Display Figure if there is only one case.  Note that the "1" below can ber changed to generate the case-by-case plots.
	if (length(caseLines) == 1) {
		figureSubSection = newSubSection("Figures");
		for(i in 1:length(caseLines)){
			caseName = caseLines[i]
			
			figureSubSection <- addTo( figureSubSection ,
					newParagraph(paste("<h3>Case ", i ,": ", caseName,"</h3>"))
			);
			
			### Create Figures
			#TODO: Low priority:  Break this into a function
			orchestraPlot_A <- paste(sep="", destFigureDir, '/', caseName,"_ArtifactMode_hist2d_FDR.png");
			figure1 <- newFigure( orchestraPlot_A, 
					"Histogram of artifact mode mutations.  All mutations to the right of the line have been filtered." );
			figureSubSection <- addTo( figureSubSection , figure1);
			
			orchestraPlot_nA <- paste(sep="", destFigureDir, '/', caseName,"_NonArtifactMode_hist2d.png");
			figure2 <- newFigure( orchestraPlot_nA,
					"Histogram of non-artifact mode mutations." );
			figureSubSection <- addTo( figureSubSection , figure2);
			
			lego_before <-  paste(sep="", destFigureDir, '/', caseName,"_lego_before.png");
			figure3 <- newFigure( lego_before,
					"Mutation spectrum before filtering." );
			figureSubSection <- addTo( figureSubSection , figure3);
			
			lego_after <-  paste(sep="", destFigureDir, '/', caseName,"_lego_after.png");
			figure4 <- newFigure( lego_after,
					"Mutation spectrum after filtering." );
			figureSubSection <- addTo( figureSubSection , figure4);
		
		}
		
		oxoReport <- addToResults(oxoReport, figureSubSection);
	}
	
	## Summary Figures
	cat(paste(sep="", "Looking for summary figures in ", destFigureDir, "\r\n"))
	rrFilename = paste(sep="",  destFigureDir, '/', basename(filteredMafFile), ".RR.png");
	paFilename = paste(sep="",  destFigureDir, '/', basename(filteredMafFile), ".PA.png");
	afFilename = paste(sep="",  destFigureDir, '/', basename(filteredMafFile), "_AF.png");
	
	figRR <- newFigure(rrFilename, "Rejected real estimates for all cases");
	oxoReport <- addToSummary(oxoReport, figRR);
	figPA <- newFigure(paFilename, "Passed artifact estimates for all cases");
	oxoReport <- addToSummary(oxoReport, figPA);
	figAF <- newFigure(afFilename, "Allelic fraction distribution of filtered mutations and unfiltered mutations.  Unfiltered mutations are counted regardless of artifact mode.  Coverage for a mutation not considered in the counts.");
	oxoReport <- addToSummary(oxoReport, figAF);

	# Do not generate the summary orchestra plots if there is only once case in the whole maf file.
	if (length(caseLines) > 1){		
		cat("Adding summary orchestra plots ...")
		figO_A <- newFigure( paste(sep="",  destFigureDir, '/', basename(filteredMafFile), " All Artifact Mode_ArtifactMode_hist2d_FDR.png"), paste(sep="","All artifact mode mutations for all ", length(caseLines)," cases.  A line (where mutations are filtered) cannot be displayed, since this varies from case to case.") );
		oxoReport <- addToSummary(oxoReport, figO_A);
		figO_nA <- newFigure( paste(sep="",  destFigureDir, '/', basename(filteredMafFile), " All Non-Artifact Mode_NonArtifactMode_hist2d.png"), paste(sep="", "All non-artifact mode mutations for all ", length(caseLines)," cases") );
		oxoReport <- addToSummary(oxoReport, figO_nA);
		figL_b <- newFigure( paste(sep="",  destFigureDir, '/', basename(filteredMafFile), "_lego_before.png"), "All mutations for this set.");
		oxoReport <- addToSummary(oxoReport, figL_b);
		figL_a <- newFigure( paste(sep="",  destFigureDir, '/', basename(filteredMafFile), "_lego_after.png"), "Filtered mutations for this set.");
		oxoReport <- addToSummary(oxoReport, figL_a);
		figH <- newFigure( paste(sep="",  destFigureDir, '/', basename(filteredMafFile), "_filteredProportion_hist.png"), paste(sep="","Case histogram of filtered mutation proportions for this set.  Number of cases: ", length(caseLines)) );
		oxoReport <- addToSummary(oxoReport, figH);
	}

	## METHODS
	oxoReport <- addToMethods( oxoReport ,
			newParagraph("<h3>Static parameters for all cases</h3>"),
			newParagraph("PoxoG -- binomial probability that a read pair from an OxoG ALT allele will be in the OxoG orientation"),
			newParagraph("PoxoG = .96"), 
			newParagraph(" "),
			newParagraph(paste(sep=" ","Passing Artifact Threshold Rate = .01", "(i.e. at most, allow an expected 1 percent of the filtered calls to be artifacts)")),
			newParagraph(paste(sep=" ","lod0 threshold = -1", "(i.e. do not filter based on lod0"))
			
	);
	
	writeReport( oxoReport , filename="nozzle", level=PROTECTION.PUBLIC );	
	
#}

